File: framework/core/src/python/kungfu/agent/action_geometry.py
# SPDX-License-Identifier: Apache-2.0

"""Versioned KFD-7 Action Geometry without adopter-domain policy.

Public evaluate operations require native ``action_runtime``. The ``*_python``
functions are explicit conformance oracles, never a silent product fallback.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from kungfu import contract as contract_runtime
from kungfu.agent.native_authority import (
    require_action_runtime,
    require_conformance_oracle,
)
from kungfu.storage import service as storage_service


SURFACE = "action-geometry"
EVALUATION_SCHEMA = "kungfu.action-geometry.evaluation/v1"
SESSION_EVALUATION_SCHEMA = "kungfu.action-geometry.session-evaluation/v1"


def contract() -> dict[str, Any]:
    return contract_runtime.load_contract(SURFACE)


def metadata() -> dict[str, str | int]:
    return contract_runtime.contract_metadata(SURFACE)


def evaluate_python(
    responsibility_ids: Mapping[str, str],
    *,
    inference_claims: Sequence[str] = (),
    conformance: bool = False,
) -> dict[str, Any]:
    """Pure-Python conformance oracle for native geometry evaluation."""

    require_conformance_oracle(conformance=conformance)
    geometry = contract()
    required = list(geometry["responsibilities"])
    supplied = set(responsibility_ids)
    required_set = set(required)
    failures: list[dict[str, Any]] = []

    missing = sorted(required_set - supplied)
    unexpected = sorted(supplied - required_set)
    if missing or unexpected:
        failures.append(
            {
                "code": "responsibility-topology-mismatch",
                "missing": missing,
                "unexpected": unexpected,
            }
        )

    identities = [
        responsibility_ids[name]
        for name in required
        if isinstance(responsibility_ids.get(name), str) and responsibility_ids[name]
    ]
    if len(identities) != len(required) or len(set(identities)) != len(identities):
        failures.append({"code": "responsibility-identity-alias"})

    forbidden = {row["forbids"]: row["id"] for row in geometry["invariants"]}
    failures.extend(
        {"code": "non-substitution-invariant", "invariant": forbidden[claim]}
        for claim in inference_claims
        if claim in forbidden
    )

    return {
        "schema": EVALUATION_SCHEMA,
        "geometryRoot": metadata()["hash"],
        "admissible": not failures,
        "responsibilityIds": {
            name: responsibility_ids[name]
            for name in required
            if name in responsibility_ids
        },
        "failures": failures,
    }


def evaluate_session_refinement_python(
    before: Mapping[str, Any],
    after: Mapping[str, Any],
    *,
    conformance: bool = False,
) -> dict[str, Any]:
    """Pure-Python conformance oracle for native session refinement."""

    require_conformance_oracle(conformance=conformance)
    geometry = contract()
    dimensions = list(geometry["sessionRefinement"]["semanticDimensions"])
    missing = [name for name in dimensions if name not in before or name not in after]
    changed = [
        name
        for name in dimensions
        if name in before and name in after and before[name] != after[name]
    ]
    return {
        "schema": SESSION_EVALUATION_SCHEMA,
        "geometryRoot": metadata()["hash"],
        "preserved": not missing and not changed,
        "missingDimensions": missing,
        "changedDimensions": changed,
    }


def evaluate(
    responsibility_ids: Mapping[str, str],
    *,
    inference_claims: Sequence[str] = (),
) -> dict[str, Any]:
    """Evaluate responsibility topology and non-substitution invariants."""

    require_action_runtime()
    return storage_service.action_runtime(
        "",
        "evaluate",
        {
            "responsibility_ids": dict(responsibility_ids),
            "inference_claims": list(inference_claims),
        },
    )


def evaluate_session_refinement(
    before: Mapping[str, Any],
    after: Mapping[str, Any],
) -> dict[str, Any]:
    """Check the geometry's conservative session round-trip dimensions."""

    require_action_runtime()
    return storage_service.action_runtime(
        "",
        "evaluate_session_refinement",
        {"before": dict(before), "after": dict(after)},
    )

File: framework/core/src/libkungfu/src/runtime/action/action_geometry.cpp
// SPDX-License-Identifier: Apache-2.0

#include <kungfu/runtime/action/action_geometry.h>

#include <kungfu/runtime/action/action_contract_registry.h>

#include <map>
#include <set>
#include <string>
#include <vector>

namespace kungfu::runtime::action {

namespace {

// Ordered responsibility names as declared by the contract.
std::vector<std::string> responsibility_order(const nlohmann::json &geometry) {
  std::vector<std::string> order;
  for (const auto &name : geometry.at("responsibilities")) {
    order.push_back(name.get<std::string>());
  }
  return order;
}

} // namespace

std::string action_geometry_root(const std::string &search_base) {
  return load_registered_contract(ACTION_GEOMETRY_SURFACE, search_base).root;
}

nlohmann::json evaluate_action_geometry(const nlohmann::json &responsibility_ids,
                                        const nlohmann::json &inference_claims, const std::string &search_base) {
  const auto contract = load_registered_contract(ACTION_GEOMETRY_SURFACE, search_base);
  const auto &geometry = contract.document;

  const auto required = responsibility_order(geometry);
  const std::set<std::string> required_set(required.begin(), required.end());

  std::set<std::string> supplied;
  if (responsibility_ids.is_object()) {
    for (const auto &item : responsibility_ids.items()) {
      supplied.insert(item.key());
    }
  }

  auto failures = nlohmann::json::array();

  // std::set iteration is ordered, matching Python's sorted(...) output.
  std::vector<std::string> missing;
  for (const auto &name : required_set) {
    if (supplied.find(name) == supplied.end()) {
      missing.push_back(name);
    }
  }
  std::vector<std::string> unexpected;
  for (const auto &name : supplied) {
    if (required_set.find(name) == required_set.end()) {
      unexpected.push_back(name);
    }
  }
  if (!missing.empty() || !unexpected.empty()) {
    failures.push_back(
        {{"code", "responsibility-topology-mismatch"}, {"missing", missing}, {"unexpected", unexpected}});
  }

  std::vector<std::string> identities;
  for (const auto &name : required) {
    if (responsibility_ids.contains(name) && responsibility_ids.at(name).is_string()) {
      auto identity = responsibility_ids.at(name).get<std::string>();
      if (!identity.empty()) {
        identities.push_back(identity);
      }
    }
  }
  const std::set<std::string> unique_identities(identities.begin(), identities.end());
  if (identities.size() != required.size() || unique_identities.size() != identities.size()) {
    failures.push_back({{"code", "responsibility-identity-alias"}});
  }

  std::map<std::string, std::string> forbidden;
  for (const auto &row : geometry.at("invariants")) {
    forbidden[row.at("forbids").get<std::string>()] = row.at("id").get<std::string>();
  }
  if (inference_claims.is_array()) {
    for (const auto &claim : inference_claims) {
      const auto key = claim.get<std::string>();
      const auto found = forbidden.find(key);
      if (found != forbidden.end()) {
        failures.push_back({{"code", "non-substitution-invariant"}, {"invariant", found->second}});
      }
    }
  }

  auto responsibility_out = nlohmann::json::object();
  for (const auto &name : required) {
    if (responsibility_ids.contains(name)) {
      responsibility_out[name] = responsibility_ids.at(name);
    }
  }

  nlohmann::json result;
  result["schema"] = ACTION_GEOMETRY_EVALUATION_V1;
  result["geometryRoot"] = contract.root;
  result["admissible"] = failures.empty();
  result["responsibilityIds"] = responsibility_out;
  result["failures"] = failures;
  return result;
}

nlohmann::json evaluate_session_refinement(const nlohmann::json &before, const nlohmann::json &after,
                                           const std::string &search_base) {
  const auto contract = load_registered_contract(ACTION_GEOMETRY_SURFACE, search_base);
  const auto &geometry = contract.document;

  std::vector<std::string> dimensions;
  for (const auto &name : geometry.at("sessionRefinement").at("semanticDimensions")) {
    dimensions.push_back(name.get<std::string>());
  }

  auto missing = nlohmann::json::array();
  for (const auto &name : dimensions) {
    if (!before.contains(name) || !after.contains(name)) {
      missing.push_back(name);
    }
  }
  auto changed = nlohmann::json::array();
  for (const auto &name : dimensions) {
    if (before.contains(name) && after.contains(name) && before.at(name) != after.at(name)) {
      changed.push_back(name);
    }
  }

  nlohmann::json result;
  result["schema"] = ACTION_GEOMETRY_SESSION_EVALUATION_V1;
  result["geometryRoot"] = contract.root;
  result["preserved"] = missing.empty() && changed.empty();
  result["missingDimensions"] = missing;
  result["changedDimensions"] = changed;
  return result;
}

} // namespace kungfu::runtime::action
